packages/ui/hooks/index.ts

```diff
@@ -74,6 +74,27 @@ export function useIsMounted() {
    return isMounted;
 }
 
+const isClient = typeof window !== 'undefined';
+let hydrationComplete = false;
+
+/**
+ * Returns true on all client-side renders *after* hydration is complete.
+ * Useful for deferring some components to the client-side while allowing later
+ * mounts to render immediately (instead of waiting for useEffect()).
+ */
+export function useIsClientSide() {
```

Comment:
I don't understand the benefit of this hook over checking `typeof window !== 'undefined'` in the Footer itself. Is the intent to let components above the fold render initially (and the footer is blank), and defer rendering the real Footer until after the first render?

Reply:
React mandates that the first render (the hydration step) produce exactly the same DOM structure that is already present in the DOM. So, the first pass (hydration) has to result in `false` even when on the client.

The `useEffect()` is run right after the hydration finishes and that changes both the `isClientSide` (triggering a re-render) and the `hydrationComplete = true`. This ensures re-renders result in `true`, and the `hydrationComplete` ensures that re-mounts also result in `true` instead of `false` before `useEffect()`.

Reply:
Thanks for the explanation! I misunderstood hydration to mean "attaching event listeners after putting the html on the page" and I hadn't considered remounting.

Reply:
> misunderstood hydration to mean "attaching event listeners after putting the html on the page"

Practically that's the end result, but it accomplishes this by performing a full render of your page, evaluating all your components, connecting existing html elements with the structure returned by the first render. If the structures don't match 100%, react complains.